slatedb/src/clock.rs

```diff
@@ -0,0 +1,246 @@
+use std::{
+    fmt::Debug,
+    sync::{
+        atomic::{AtomicI64, Ordering},
+        Arc,
+    },
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use tracing::info;
+
+use crate::{config::DurabilityLevel, SlateDBError};
+
+/// Defines the generic clock trait that SlateDB extends for user and system operations.
+pub trait Clock: Debug + Send + Sync {
+    /// Returns a timestamp (typically measured in millis since the unix epoch),
+    /// must return monotonically increasing numbers (this is enforced
+    /// at runtime and will panic if the invariant is broken).
+    ///
+    /// Note that this clock does not need to return a number that
+    /// represents the unix timestamp; the only requirement is that
+    /// it represents a sequence that can attribute a logical ordering
+    /// to actions on the database.
+    fn now(&self) -> i64;
+}
+
+/// Defines the clock that SlateDB will use for user facing operations.
+pub trait UserClock: Clock {
+    /// Returns the current time plus the given duration.
+    fn add(&self, duration: Duration) -> i64 {
+        self.now() + duration.as_millis() as i64
+    }
+}
+
+/// Defines the clock that SlateDB will use for system operations.
+pub trait SystemClock: Clock {
+    /// Returns the current time as a SystemTime.
+    ///
+    /// This function panics if the Clock time cannot be converted to a SystemTime.
+    fn now_systime(&self) -> SystemTime {
+        chrono::DateTime::from_timestamp_millis(self.now())
+            .map(SystemTime::from)
+            .expect("Failed to convert Clock time to SystemTime")
+    }
+
+    /// Returns the time that has elapsed since the unix epoch.
+    ///
+    /// This function returns `0` if the current time is before the unix epoch.
+    fn elapsed(&self) -> Duration {
+        self.now_systime()
+            .duration_since(UNIX_EPOCH)
+            .unwrap_or_default()
+    }
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+#[derive(Debug)]
+pub struct DefaultSystemClock {
+    last_tick: AtomicI64,
+}
+
+impl Default for DefaultSystemClock {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DefaultSystemClock {
+    pub fn new() -> Self {
+        Self {
+            last_tick: AtomicI64::new(i64::MIN),
+        }
+    }
+}
+
+impl Clock for DefaultSystemClock {
+    fn now(&self) -> i64 {
+        // since SystemTime is not guaranteed to be monotonic, we enforce it here
+        let tick = match SystemTime::now().duration_since(UNIX_EPOCH) {
+            Ok(duration) => duration.as_millis() as i64, // Time is after the epoch
+            Err(e) => -(e.duration().as_millis() as i64), // Time is before the epoch, return negative
+        };
+        self.last_tick.fetch_max(tick, Ordering::SeqCst).max(tick)
+    }
+}
+
+impl UserClock for DefaultSystemClock {}
```

Comment:
I thought this might defeat the purpose of splitting the types. After some consideration, I think it's actually OK, though. My initial concern was that you could do:

```rust
let system_clock = DefaultSystemClock::new();
//...
    .with_user_clock(system_clock)
```

In realistic scenarios, I think you'd probably not be calling `with_*_clock` using `DefaultSystemClock` anyway. Instead, you'd probably have:

```rust
// MyUserClock implements UserClock but not SystemClock traits
let user_clock = MyUserClock::new();
//...
    .with_user_clock(user_clock);
```

In such a case, if `with_system_clock` is called, it will fail at compile time.

So I think the type hierarchy seems right even though DefaultSystemClock can be used for both `with`'s.

Reply:
I'm a bit confused about the type hierarchy and why we need to share types at all? it seems to me like two totally distinct concepts: there's a user clock and there's a system clock. It seems fine to me if they're two different types that don't (and can't) overlap. Maybe I'm missing a use case here though?

I don't see, for example, why the system clock ever needs to return `now() -> i64` and then also have `now_systime`.